community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCacheTest.java

```diff
@@ -214,6 +220,247 @@ public void mustFlushDirtyPagesOnEvictingAllPages() throws Exception
+        MuninnPageCache pageCache =
+                createPageCache( fs, 2, 8, PageCacheTracer.NULL, PageCursorTracerSupplier.NULL, versionContextSupplier );
+
+        PagedFile pagedFile = pageCache.map( file( "a" ), 8 );
+        cursorContext.initWrite( 7 );
+        try ( PageCursor cursor = pagedFile.io( 0, PF_SHARED_WRITE_LOCK ) )
+        {
+            assertTrue( cursor.next() );
+            cursor.putLong( 3 );
+        }
+
+        cursorContext.initRead();
+        assertFalse( cursorContext.isDirty() );
+        try ( PageCursor cursor = pagedFile.io( 0, PF_SHARED_READ_LOCK ) )
+        {
+            assertTrue( cursor.next() );
+            assertEquals( 3, cursor.getLong() );
+            assertTrue( cursorContext.isDirty() );
+        }
+    }
+
+    @Test
+    public void doNotMarkCursorContextAsDirtyWhenReadingDataFromOlderTransactions() throws IOException
+    {
+        TestVersionContext cursorContext = new TestVersionContext( () -> 23 );
+        VersionContextSupplier versionContextSupplier = new ConfiguredVersionContextSupplier( cursorContext );
+        MuninnPageCache pageCache =
+                createPageCache( fs, 2, 8, PageCacheTracer.NULL, PageCursorTracerSupplier.NULL, versionContextSupplier );
+
+        PagedFile pagedFile = pageCache.map( file( "a" ), 8 );
+        cursorContext.initWrite( 17 );
+        try ( PageCursor cursor = pagedFile.io( 0, PF_SHARED_WRITE_LOCK ) )
+        {
+            assertTrue( cursor.next() );
+            cursor.putLong( 3 );
+        }
+
+        cursorContext.initRead();
+        assertFalse( cursorContext.isDirty() );
+        try ( PageCursor cursor = pagedFile.io( 0, PF_SHARED_READ_LOCK ) )
+        {
+            assertTrue( cursor.next() );
+            assertEquals( 3, cursor.getLong() );
+            assertFalse( cursorContext.isDirty() );
+        }
+    }
+
+    @Test
+    public void markContextAsDirtyWhenAnyEvictedPageHaveModificationTransactionHigherThenReader() throws IOException
+    {
+        TestVersionContext cursorContext = new TestVersionContext( () -> 5 );
+        VersionContextSupplier versionContextSupplier = new ConfiguredVersionContextSupplier( cursorContext );
+        MuninnPageCache pageCache =
+                createPageCache( fs, 2, 8, PageCacheTracer.NULL, PageCursorTracerSupplier.NULL, versionContextSupplier );
+
+        PagedFile pagedFile = pageCache.map( file( "a" ), 8 );
+        cursorContext.initWrite( 3 );
+        try ( PageCursor cursor = pagedFile.io( 0, PF_SHARED_WRITE_LOCK ) )
+        {
+            assertTrue( cursor.next() );
+            cursor.putLong( 3 );
+        }
+
+        cursorContext.initWrite( 13 );
+        try ( PageCursor cursor = pagedFile.io( 1, PF_SHARED_WRITE_LOCK ) )
+        {
+            assertTrue( cursor.next() );
+            cursor.putLong( 4 );
+        }
+
+        pageCache.pages.tryEvict( pageCache.grabFreeAndExclusivelyLockedPage( PageFaultEvent.NULL ),
+                EvictionRunEvent.NULL );
```

Comment:
Resurfacing comment: It doesn't make sense to try to evict a free page. The `tryEvict` will always return `false` because the given page is already exclusively locked. The `grabFreeAndExclusivelyLockedPage` method might do a cooperative eviction, but the evicted page will be chosen at random. Can we make this test deterministic?